Add Too Many Constants smell detector

waldyr opened the pull request:
#621 

Even thought this might still be in discussion I went ahead and implemented. It may help in the discussion. What do you guys think?

There's one spec that I've skipped because I didn't know what would be the correct context to inspect.

I used 5 as max quantity because I didn't want to fix [nested_iterators](https://github.com/troessner/reek/blob/master/lib/reek/smells/nested_iterators.rb) 😂 

Also I want to add Exception Handling into the discussion because It's pretty common to find code like this:
```
MyException = Class.new(StandardError)
```

mvz commented on spec/reek/smells/too_many_constants_spec.rb:
```diff
@@ -0,0 +1,125 @@
+    end
+  end
+
+  context 'smell report' do
```
You can combine these three expectations into one.

mvz replied:
Alternatively, you can at least pull the definition of `src` out into a `let` statement.

waldyr replied:
Not a `let`'s fan but I'll improve this expectations

mvz commented on spec/reek/smells/too_many_constants_spec.rb:
```diff
@@ -0,0 +1,125 @@
+    it 'should not report for non-excessive constants' do
+      src = <<-EOS
+        class Dummy
+          #{casgn_sequence}
```
Having this indirection makes the test less readable I think, since it is not immediately clear how many constants are created here. Just writing it all out makes the tests a bit longer but a lot clearer.

mvz commented on spec/reek/smells/too_many_constants_spec.rb:
```diff
@@ -0,0 +1,125 @@
+      expect(src).to reek_of(:TooManyConstants)
+    end
+
+    xit 'should report for excessive constants anywhere' do
```
This case is outside of any of the contexts you're checking. What *may* work is adding `:root` to the list of contexts, but I don't think we check within that context in any of the other detectors, so you'll be a bit in uncharted territory there.

waldyr replied:
I will need at least an if clause to treat things that is only inside the `:root` context. Feasible, but viable?

mvz commented on lib/reek/smells/too_many_constants.rb:
```diff
@@ -0,0 +1,52 @@
+      def inspect(ctx)
+        max_allowed_consts = value(MAX_ALLOWED_CONSTS_KEY, ctx)
+
+        count = ctx.each_node(:casgn).map! { |casgn| casgn.children.last }.uniq.length
```
This looks very complex when I compare it to the specs.

mvz commented on spec/reek/smells/too_many_constants_spec.rb:
```diff
@@ -0,0 +1,125 @@
+        end
+      EOS
+
+      expect(src).to reek_of(:TooManyConstants, message: "has at least #{too_many_consts} constants")
```
Be sure to also check at that it reports on the context `Moduly::Class`, and not on the context `Moduly`.

mvz commented:
> Also I want to add Exception Handling into the discussion because It's pretty common to find code like this:

> ```
> MyException = Class.new(StandardError)
> ```

I think that code should count as a class definition and not as a constant in the regular sense.

troessner commented:
Gonna review this bad boy as well in the next days!
@waldyr I just added you as "collaborateur" to this repo so no more need for forking the repo, you can create directly pull requests ;)

waldyr commented on lib/reek/smells/too_many_constants.rb:
```diff
@@ -0,0 +1,52 @@
+      def inspect(ctx)
+        max_allowed_consts = value(MAX_ALLOWED_CONSTS_KEY, ctx)
+
+        count = ctx.each_node(:casgn).delete_if(&:constant_definition).length
```
This makes me want to change the [`constant_definition`](https://github.com/waldyr/reek/blob/master/lib/reek/ast/sexp_extensions/module.rb#L90)

troessner commented on defaults.reek:
```diff
@@ -115,3 +119,4 @@ UnusedPrivateMethod:
 UtilityFunction:
   enabled: true
   exclude: []
+  public_methods_only: false
```
This doesnt belong here but should go away after a rebase.

troessner commented on docs/Too-Many-Constants.md:
````diff
@@ -0,0 +1,37 @@
+
+```
+test.rb -- 5 warnings:
+  [1]:TooManyConstants: Smelly has at least 4 instance variables
````
Instance variables? C&p error?

troessner replied:
Also given the configuration above Reek would **not** emit any warnings.

troessner commented on docs/Too-Many-Constants.md:
````diff
@@ -0,0 +1,37 @@
+Reek would emit the following warning:
+
+```
+test.rb -- 5 warnings:
````
Lets redact this to say "1", otherwise it is confusing for the reader without seeing all the other  warnings.

troessner commented on features/smells/too_many_constants.feature:
```diff
@@ -0,0 +1,19 @@
+    Then it reports:
+    """
+    my_class.rb -- 1 warning:
+      [1]:TooManyConstants: MyClass has at least 6 constants [https://github.com/troessner/reek/blob/master/docs/Too-Many-Constants.md]
```
"At least"? That is confusing - why not "has X constants"?

troessner commented on lib/reek/report/code_climate/code_climate_configuration.yml:
````diff
@@ -525,6 +525,37 @@ TooManyInstanceVariables:
+
+    ```
+    test.rb -- 5 warnings:
+      [1]:TooManyConstants has at least 4 constants (TooManyConstants)
````
Same here, lets redact the number.

troessner commented on lib/reek/smells/too_many_constants.rb:
```diff
@@ -0,0 +1,51 @@
+    class TooManyConstants < SmellDetector
+      # The name of the config field that sets the maximum number
+      # of constants permitted in a class.
+      MAX_ALLOWED_CONSTS_KEY = 'max_constants'.freeze
```
We dont abbreviate this anywhere so lets rather spell it out....

troessner commented on lib/reek/smells/too_many_constants.rb:
```diff
@@ -0,0 +1,51 @@
+
+      def self.contexts
+        [:class, :module]
+      end
```
Maybe worth investigating: i think "module" covery class as well.

waldyr replied:
If a class is inside the module it will cover. However we might have classes without namespace like:
```
class Test
end
```
And in this case we should look into class context.

mvz commented on spec/reek/smells/too_many_constants_spec.rb:
```diff
@@ -0,0 +1,115 @@
+      expect(src).to reek_of(:TooManyConstants, lines: [1])
+    end
+
+    it 'reports the line where it occurs' do
```
-> reports the module.

mvz commented on spec/reek/smells/too_many_constants_spec.rb:
```diff
@@ -0,0 +1,115 @@
+    end
+
+    it 'reports the line where it occurs' do
+      expect(src).to reek_of(:TooManyConstants, context: 'Moduly::Klass')
```
Please also check no smell is reported on Moduly.

waldyr replied:
As the class is also inside module both the class and the module is going to smell.

mvz commented on spec/reek/smells/too_many_constants_spec.rb:
```diff
@@ -0,0 +1,129 @@
+    end
+
+    it 'should count each constant only once' do
+      src = <<-EOS
```
If you put these three classes together in a module, my bet is the module will smell. I don't think this is what we want.

waldyr replied:
Got it.

mvz commented on spec/reek/smells/too_many_constants_spec.rb:
```diff
@@ -0,0 +1,129 @@
+    end
+
+    it 'reports the context where it occurs' do
+      expect(src).to reek_of(:TooManyConstants, context: 'Moduly::Klass')
```
Only `Moduly::Klass` should smell, not `Moduly`.

waldyr commented on spec/reek/smells/too_many_constants_spec.rb:
```diff
@@ -0,0 +1,174 @@
+    end
+  end
+
+  context 'smell report' do
```
@mvz I'd rather repeat it inside the `it` block as many times as needed instead of using this approach or `let`. The reason behind it is even though repeating it goes against DRY principle, in an `it` block I will have everything that is required to understand what is being tested and how.

For instance, let's suppose we start irresponsibly adding more `context`s and `it`s blocks inside this context. Even though I know there's a `src` it will become more and more difficult to find where it is defined. In every line addition the `src` variable will turn into a even more mysterious guest.

Even though the big block of text I'm fine with any possible conclusion. What do you think?

waldyr commented:
@mvz @troessner 

What about this one? Am I missing something here?

mvz commented:
Oooh ... this one totally went off the radar. I'll try to have a look at it this weekend.

troessner commented on lib/reek/smells/too_many_constants.rb:
```diff
@@ -0,0 +1,57 @@
+      # The name of the config field that sets the maximum number
+      # of constants permitted in a class.
+      MAX_ALLOWED_CONSTANTS_KEY = 'max_constants'.freeze
+      DEFAULT_MAX_CONSTS = 5
```
"Constants" is nowhere abbreviated so let's use "DEFAULT_MAX_CONSTANTS" here.

troessner commented on spec/reek/smells/too_many_constants_spec.rb:
```diff
@@ -0,0 +1,216 @@
+      expect(src).not_to reek_of(:TooManyConstants)
+    end
+
+    it 'should not report when increasing default' do
```
Nice spec!

waldyr replied:
Props to mutant. I really tried to make Travis green.

troessner commented on spec/reek/smells/too_many_constants_spec.rb:
```diff
@@ -0,0 +1,216 @@
+  end
+
+  context 'smell report' do
+    it 'reports the number of consts' do
```
"constants" please everywhere.

troessner commented on lib/reek/smells/too_many_constants.rb:
```diff
@@ -0,0 +1,57 @@
+      # @return [Array<SmellWarning>]
+      #
+      def inspect(ctx)
+        max_allowed_consts = value(MAX_ALLOWED_CONSTANTS_KEY, ctx)
```
And here :)

troessner commented:
Couple of small things and travis as usual :)
Then LGTM and put a stop to further detectors until we refactored our specs.

waldyr commented:
@troessner 

> Couple of small things and travis as usual :)

Addressed!

> Then LGTM and put a stop to further detectors until we refactored our specs.

Okay. Going to jump in the discussion.

troessner commented:
Ok, the remaining mutant failures are "ok" I guess. This PR is good to go.
Awesome job @waldyr !

troessner commented:
And I just released 4.2.0!
